core/src/test/java/io/grpc/PickFirstLoadBalancer2Test.java

```diff
@@ -0,0 +1,232 @@
+    inOrder.verify(mockHelper).updatePicker(pickerCaptor.capture());
+    assertEquals(socketAddresses, eagCaptor.getValue().getAddresses());
+
+    loadBalancer.handleResolvedAddresses(newServers, affinity);
+    inOrder.verify(mockHelper).createSubchannel(eagCaptor.capture(), any(Attributes.class));
+    inOrder.verify(mockHelper).updatePicker(pickerCaptor.capture());
+    assertEquals(newSocketAddresses, eagCaptor.getValue().getAddresses());
+
+    Subchannel subchannel = pickerCaptor.getAllValues().get(0).pickSubchannel(
+        affinity, new Metadata()).getSubchannel();
+    assertEquals(oldSubchannel, subchannel);
+
+    Subchannel subchannel2 = pickerCaptor.getAllValues().get(1).pickSubchannel(affinity,
+        new Metadata()).getSubchannel();
+    assertEquals(newSubchannel, subchannel2);
+    verify(subchannel2, never()).shutdown();
+
+    verifyNoMoreInteractions(mockHelper);
+  }
+
+  @Test
+  public void stateChangeBeforeResolution() throws Exception {
+    loadBalancer.handleSubchannelState(mockSubchannel,
+        ConnectivityStateInfo.forNonError(ConnectivityState.READY));
+
+    verifyNoMoreInteractions(mockHelper);
+  }
+
+  @Test
+  public void pickAfterStateChangeAfterResolution() throws Exception {
+    loadBalancer.handleResolvedAddresses(servers, affinity);
+    verify(mockHelper).updatePicker(pickerCaptor.capture());
+    Subchannel subchannel = pickerCaptor.getValue().pickSubchannel(affinity,
+        new Metadata()).getSubchannel();
+    reset(mockHelper);
+
+    loadBalancer.handleSubchannelState(subchannel,
+        ConnectivityStateInfo.forTransientFailure(Status.UNAVAILABLE));
+    loadBalancer.handleSubchannelState(subchannel,
+        ConnectivityStateInfo.forNonError(ConnectivityState.IDLE));
+    loadBalancer.handleSubchannelState(subchannel,
+        ConnectivityStateInfo.forNonError(ConnectivityState.READY));
+
+    verify(mockHelper, times(3)).updatePicker(pickerCaptor.capture());
+
+    assertEquals(Status.UNAVAILABLE,
+        pickerCaptor.getAllValues().get(1).pickSubchannel(Attributes.EMPTY,
+            new Metadata()).getStatus());
+    assertEquals(Status.OK, pickerCaptor.getAllValues().get(2).pickSubchannel(Attributes.EMPTY,
+        new Metadata()).getStatus());
+    assertEquals(PickResult.withSubchannel(subchannel).getSubchannel(), pickerCaptor.getAllValues()
+        .get(3).pickSubchannel(Attributes.EMPTY, new Metadata()).getSubchannel());
+
+    verifyNoMoreInteractions(mockHelper);
+  }
+
+  @Test
+  public void nameResolutionError() throws Exception {
+    loadBalancer.handleNameResolutionError(Status.UNKNOWN);
+    verify(mockHelper).updatePicker(pickerCaptor.capture());
+    PickResult pickResult = pickerCaptor.getValue().pickSubchannel(Attributes.EMPTY,
+        new Metadata());
+    assertEquals(null, pickResult.getSubchannel());
+    assertEquals(Status.UNKNOWN, pickResult.getStatus());
+    verifyNoMoreInteractions(mockHelper);
+  }
+
```

Comment:
Unclear to me what value this will provide unless you're implying that we should change `handleNameResolutionError()` implementation which is currently unconditional and very simple: 

```
    public void handleNameResolutionError(Status error) {
      helper.updatePicker(new Picker(PickResult.withError(error)));
    }
```

so no matter what happens before/after, it'll *always* just propagate resolution error down to `updatePicker`.